engine/common/follower/pending_tree/pending_tree.go

```diff
@@ -0,0 +1,198 @@
+				// this vertex is already in tree, skip it
+				continue
+			} else {
+				return nil, model.ByzantineThresholdExceededError{Evidence: fmt.Sprintf(
+					"conflicting QCs at view %d: %v and %v",
+					block.View(), v.ID(), block.ID(),
+				)}
+			}
+		}
+
+		vertex, err := NewVertex(block, false)
+		if err != nil {
+			return nil, fmt.Errorf("could not create new vertex: %w", err)
+		}
+		err = t.forest.VerifyVertex(vertex)
+		if err != nil {
+			return nil, fmt.Errorf("failed to store certified block into the tree: %w", err)
+		}
+		t.forest.AddVertex(vertex)
+
+		if t.connectsToFinalizedBlock(block) {
+			connectedBlocks := t.updateAndCollectFork(vertex)
+			allConnectedBlocks = append(allConnectedBlocks, connectedBlocks...)
+		}
+	}
+
+	return allConnectedBlocks, nil
+}
+
+// connectsToFinalizedBlock checks if candidate block connects to the finalized state.
+func (t *PendingTree) connectsToFinalizedBlock(block CertifiedBlock) bool {
+	if block.Block.Header.ParentID == t.lastFinalizedID {
+		return true
+	} else if parentVertex, found := t.forest.GetVertex(block.Block.Header.ParentID); found {
+		return parentVertex.(*PendingBlockVertex).connectedToFinalized
+	} else {
+		return false
+	}
+}
+
+// FinalizeForkAtLevel takes last finalized block and prunes levels below the finalized view.
+// When a block is finalized we don't care for all blocks below it since they were already finalized.
+// No errors are expected during normal operation.
+func (t *PendingTree) FinalizeForkAtLevel(finalized *flow.Header) error {
+	blockID := finalized.ID()
+	if t.forest.LowestLevel >= finalized.View {
+		return nil
+	}
+
+	t.lastFinalizedID = blockID
+	err := t.forest.PruneUpToLevel(finalized.View)
+	if err != nil {
+		return fmt.Errorf("could not prune tree up to view %d: %w", finalized.View, err)
+	}
+	return nil
+}
+
+// updateAndCollectFork marks the subtree rooted at `vertex.Block` as connected to the finalized state
+// and returns all blocks in this subtree. No parents of `vertex.Block` are modified or included in the output.
+// The output list will be ordered so that parents appear before children.
+// The caller must ensure that `vertex.Block` is connected to the finalized state.
+//
+//	A ← B ← C ←D
+//	      ↖ E
+//
+// For example, suppose B is the input vertex. Then:
+//   - A must already be connected to the finalized state
+//   - B, E, C, D are marked as connected to the finalized state and included in the output list
+func (t *PendingTree) updateAndCollectFork(vertex *PendingBlockVertex) []CertifiedBlock {
+	certifiedBlocks := []CertifiedBlock{vertex.CertifiedBlock}
+	vertex.connectedToFinalized = true
+	iter := t.forest.GetChildren(vertex.VertexID())
+	for iter.HasNext() {
+		nextVertex := iter.NextVertex().(*PendingBlockVertex)
+		// if it's already connected then it was already reported
+		if !nextVertex.connectedToFinalized {
+			blocks := t.updateAndCollectFork(nextVertex)
+			certifiedBlocks = append(certifiedBlocks, blocks...)
```

Comment:
This is very inefficiently implemented and results on O(2) memory allocations.  This is because you pre-pend, which here, involves copying the entire content of blocks. So if the recursion goes through few iterations along a fork with some depth, this can become expensive.

I would suggest to consider go's append method and do something similar:
```golang
slice = append(slice, newElements...)
```
you pass the `slice` as an input and append has the freedom to mutate the slice in place or allocate a new one. Using this pattern for your `updateAndCollectFork` method would be something like
```golang
// This method has a similar signature as `append` for performance reasons:
//  * any connected certified blocks are appended to `queue`
//  * we return the _resulting slice_ after all appends
func (t *PendingTree) updateAndCollectFork(queue []CertifiedBlock, vertex *PendingBlockVertex) []CertifiedBlock {
	vertex.connectedToFinalized = true
	queue = append(queue, vertex.CertifiedBlock)

	iter := t.forest.GetChildren(vertex.VertexID())
	for iter.HasNext() {
		nextVertex := iter.NextVertex().(*PendingBlockVertex)
		// if it's already connected then it was already reported
		if !nextVertex.connectedToFinalized {
			queue = t.updateAndCollectFork(queue, nextVertex)
		}
	}
	return queue
}
```

Reply:
Yeah, that makes sense, I will update